File: analytics/src/main/java/io/appmetrica/analytics/impl/TaskProcessor.java
package io.appmetrica.analytics.impl;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import io.appmetrica.analytics.impl.component.IComponent;
import io.appmetrica.analytics.impl.events.EventsFlusher;
import io.appmetrica.analytics.impl.startup.executor.StartupExecutor;
import io.appmetrica.analytics.logger.appmetrica.internal.DebugLogger;
import io.appmetrica.analytics.networktasks.internal.NetworkTask;

public class TaskProcessor<C extends IComponent> implements ServiceLifecycleObserver, EventsFlusher {

    private static final String TAG = "[TaskProcessor]";

    @NonNull
    private C mComponent;

    final Object mSync = new Object();

    boolean mIsShuttingDown = false;

    @NonNull
    private final StartupExecutor mStartupExecutor;

    public TaskProcessor(@NonNull C component, @NonNull StartupExecutor executor) {
        mComponent = component;
        mStartupExecutor = executor;
    }

    @Override
    public void onCreate() {
        synchronized (mSync) {
            if (mIsShuttingDown) {
                mIsShuttingDown = false;
                DebugLogger.INSTANCE.info(TAG, "start");
            }
        }
    }

    @Override
    public void onDestroy() {
        synchronized (mSync) {
            if (!mIsShuttingDown) {
                cancelFlushTask();

                mIsShuttingDown = true;

                DebugLogger.INSTANCE.info(TAG, "stop");
            }
        }
    }

    void cancelFlushTask() {

    }

    @Override
    public void flushAllTasks() {
        DebugLogger.INSTANCE.info(TAG, "flushAllTasks");
        synchronized (mSync) {
            if (!mIsShuttingDown) {
                runAllTasks();
                cancelFlushTask();
            }
        }
    }

    void runAllTasks() {
        synchronized (mSync) {
            if (!mIsShuttingDown) {
                runTasks();
            }
        }
    }

    void startTask(@NonNull NetworkTask networkTask) {
        GlobalServiceLocator.getInstance().getNetworkCore().startTask(networkTask);
    }

    void runTasks() {
        DebugLogger.INSTANCE.info(TAG, "sendStartupIfRequired");
        mStartupExecutor.sendStartupIfRequired();
    }

    @NonNull
   public C getComponent() {
        return mComponent;
    }

    @NonNull
    @VisibleForTesting
    public StartupExecutor getStartupExecutor() {
        return mStartupExecutor;
    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/TaskProcessor.kt
package io.appmetrica.analytics.impl

import io.appmetrica.analytics.impl.component.IComponent
import io.appmetrica.analytics.impl.events.EventsFlusher
import io.appmetrica.analytics.impl.startup.executor.StartupExecutor
import io.appmetrica.analytics.logger.appmetrica.internal.DebugLogger
import io.appmetrica.analytics.networktasks.internal.NetworkTask
import java.util.concurrent.atomic.AtomicBoolean

internal open class TaskProcessor<C : IComponent>(
    val component: C,
    private val mStartupExecutor: StartupExecutor
) : ServiceLifecycleObserver, EventsFlusher {
    private val shuttingDown = AtomicBoolean(false)

    private val tag = "[TaskProcessor-${component.getComponentId()}]"

    override fun onCreate() {
        if (shuttingDown.compareAndSet(true, false)) {
            DebugLogger.info(tag, "start")
        }
    }

    override fun onDestroy() {
        if (shuttingDown.compareAndSet(false, true)) {
            cancelFlushTask()
            DebugLogger.info(tag, "stop")
        }
    }

    open fun cancelFlushTask() {
    }

    override fun flushAllTaskAsync() {
        if (shuttingDown.get().not()) {
            DebugLogger.info(tag, "scheduleFlushTask")
            scheduleFlushTaskNow()
        }
    }

    open fun scheduleFlushTaskNow() {
    }

    override fun flushAllTasks() {
        DebugLogger.info(tag, "flushAllTasks")
        if (shuttingDown.get().not()) {
            runTasks()
            cancelFlushTask()
        }
    }

    fun startTask(networkTask: NetworkTask) {
        GlobalServiceLocator.getInstance().networkCore.startTask(networkTask)
    }

    open fun runTasks() {
        DebugLogger.info(tag, "sendStartupIfRequired")
        mStartupExecutor.sendStartupIfRequired()
    }

    fun isShuttingDown(): Boolean {
        return shuttingDown.get()
    }
}
